File: Loenn/entities/firewallBarrier.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local firewallBarrier = {}

firewallBarrier.name = "MaggyHelper/FirewallBarrier"
firewallBarrier.depth = -9000
firewallBarrier.minimumSize = {8, 8}
firewallBarrier.fieldInformation = {
    maxEnergy = {
        minimumValue = 0
    },
    energyDrain = {
        minimumValue = 0
    }
}
firewallBarrier.placements = {
    {
        name = "active",
        data = {
            width = 32,
            height = 32,
            maxEnergy = 100,
            energyDrain = 5,
            startActive = true
        }
    },
    {
        name = "inactive",
        data = {
            width = 32,
            height = 32,
            maxEnergy = 100,
            energyDrain = 5,
            startActive = false
        }
    }
}

function firewallBarrier.sprite(room, entity)
    local sprites = {}
    local width = entity.width or 32
    local height = entity.height or 32
    local state = entity.startActive and "active" or "inactive"
    
    for x = 0, width - 8, 8 do
        for y = 0, height - 8, 8 do
            local sprite = drawableSprite.fromTexture("objects/firewall_barrier/" .. state, entity)
            sprite:setPosition(entity.x + x, entity.y + y)
            sprite:setJustification(0, 0)
            if entity.startActive then
                sprite.color = {1.0, 0.5, 0.0, 0.8}
            end
            table.insert(sprites, sprite)
        end
    end
    
    return sprites
end

function firewallBarrier.selection(room, entity)
    return utils.rectangle(entity.x, entity.y, entity.width or 32, entity.height or 32)
end

return firewallBarrier

File: Ahorn/entities/FirewallBarrier.jl
module FirewallBarrier

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/FirewallBarrier" FirewallBarrier(x::Integer, y::Integer, energyDrain::Integer=5, height::Integer=32, maxEnergy::Integer=100, startActive::Bool=true, width::Integer=32)

const placements = Ahorn.PlacementDict(
    "active" => Ahorn.EntityPlacement(FirewallBarrier),
    "inactive" => Ahorn.EntityPlacement(FirewallBarrier)
)

function Ahorn.selection(entity::FirewallBarrier)
    return Ahorn.getEntityRectangle(entity)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::FirewallBarrier, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.getEntityRectangle(entity), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end
